File: frontend/src/api/edge.ts
import { apiGet, apiPost, apiPut, apiDelete } from './http'

export interface EdgeCreate {
  topologyId: string
  edgeTypeId: string
  sourceId: string
  targetId: string
  status?: string
}

export interface EdgeUpdate {
  status?: string | null
}

export interface EdgeDetail {
  id: string
  topologyId: string
  edgeTypeId: string
  sourceId: string
  targetId: string
  status: string
  createdAt: string
  updatedAt: string
  attrs?: Record<string, string | null>
}

export interface EdgeItem {
  id: string
  topologyId: string
  edgeTypeId: string
  sourceId: string
  targetId: string
  status: string
  createdAt: string
  updatedAt: string
}

export interface EdgeListResponse {
  items: EdgeItem[]
  total: number
  page: number
  pageSize: number
}

export const edgeApi = {
  list: (topologyId: string, params?: { edgeTypeId?: string; sourceId?: string; targetId?: string; page?: number; pageSize?: number }): Promise<EdgeListResponse> =>
    apiGet(`/topologies/${topologyId}/edges`, params),

  get: (id: string): Promise<EdgeDetail> =>
    apiGet(`/edges/${id}`),

  create: (topologyId: string, data: EdgeCreate): Promise<{ id: string }> =>
    apiPost(`/topologies/${topologyId}/edges`, data),

  update: (id: string, data: EdgeUpdate): Promise<{ id: string }> =>
    apiPut(`/edges/${id}`, data),

  delete: (id: string): Promise<{ id: string }> =>
    apiDelete(`/edges/${id}`),

  setAttrs: (id: string, attrs: Record<string, string | null>): Promise<{ id: string }> => {
    const attrsList = Object.entries(attrs).map(([field_key, value]) => ({ field_key, value }))
    return apiPut(`/edges/${id}/attrs`, attrsList)
  },
}

File: backend/app/admin/schemas/edge.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field

from ._base import CamelModel


class EdgeCreate(CamelModel):
    edge_type_id: str
    source_id: str
    target_id: str
    status: str = Field(default="up", max_length=20)


class EdgeUpdate(CamelModel):
    status: Optional[str] = Field(default=None, max_length=20)


class EdgeAttrSet(CamelModel):
    field_key: str
    value: Optional[str] = None


class EdgeItem(CamelModel):
    id: str
    topology_id: str
    edge_type_id: str
    source_id: str
    target_id: str
    status: str
    created_at: datetime
    updated_at: datetime


class EdgeDetail(EdgeItem):
    attrs: dict[str, Any] = {}


class EdgeListResponse(CamelModel):
    code: int = 0
    data: dict[str, Any]
    message: str = "ok"
